src/librustc_trans/back/link.rs

```diff
@@ -150,14 +149,35 @@ pub fn get_linker(sess: &Session) -> (String, Command, Option<PathBuf>) {
     if let Some(ref linker) = sess.opts.cg.linker {
         (linker.clone(), Command::new(linker), None)
     } else if sess.target.target.options.is_like_msvc {
-        let (cmd, host) = msvc::link_exe_cmd(sess);
+        let (cmd, host) = msvc_link_exe_cmd(sess);
         ("link.exe".to_string(), cmd, host)
     } else {
         (sess.target.target.options.linker.clone(),
          Command::new(&sess.target.target.options.linker), None)
     }
 }
 
+#[cfg(windows)]
+pub fn msvc_link_exe_cmd(sess: &Session) -> (Command, Option<PathBuf>) {
+    use gcc::windows_registry;
+
+    let target = &sess.opts.target_triple;
+    let tool = windows_registry::find_tool(target, "link.exe");
+
+    if let Some(tool) = tool {
+        let path = tool.path().parent().map(|p| p.to_owned());
+        (tool.to_command(), path)
```

Comment:
The issue is right here. `path` is *not* supposed to be the directory with the linker, but rather the directory of the *host toolchain*. You'll have to teach `gcc` how to provide that information so you can pull it out here.